### DISCUSSION TITLE
FR: Add dependencies argument to useMachine to force new service

### DISCUSSION THREAD

**Discussion by FokkeZB** [statelyai/xstate#1399] [General] [answered]
# Scenario

We have 2 screens:

- Select/Connect App Auth A
- Select/Connect App Auth B

They use the same controller component, which uses `useMachine` for state management internally.

When user advances from the first to the second screen, React re-uses the component instance and thus service returned by `useMachine`, as internally it will only create the service once:

https://github.com/davidkpiano/xstate/blob/1212d0b9a811fce98f7d369e214717bc0946d462/packages/xstate-react/src/useMachine.ts#L138-L161

As a result, the second screen will open in the same state as the first.

# Request

Support a third, `deps` argument, similar to `useMemo` that will force a new service to be created if they change.

**Comment by Andarist** [comment, marked as answer]
Can't you just use `key` prop on your component? Or push the machine to a component responsible for the not-shared screen?

**Reply by davidkpiano** [reply]
Right, this sounds like it needs a React solution, not necessarily a `useMachine` solution:

```jsx
// Force-create new instance
<MyComponent screen={currentScreen} key={currentScreen.id} />
```

**Reply by FokkeZB** [reply]
I agree that's one way to solve it (and what I currently use), but I'd prefer to encapsulate that within the component itself.

In similar fashion, I wonder how I could update a machine's action, e.g. one that fetches from an API using a token that has refreshed, *without* forcing a new service (and then having to hydrate the previous service its state).

**Reply by Andarist** [reply]
This is problematic because I think we'd like to keep `useMachine(config, options)` API. There are 2 problems with this:
- when adding `inputs` parameter we should require it to be always provided, when in most cases it would force people to pass `[]`. Technically, of course, we could just assume `[]` if not given explicitly, but mirroring React's builtin hooks and the overall community conventions seem important
- we need to cache internally created machines and `useMemo` is not really a semantic guarantee, so we'd probably have to replicate `inputs` comparisons manually in 
- currently `options` are optional, so there would be times when people would have to write `useMachine(config, undefined, [foo])` which doesn't look that good

> In similar fashion, I wonder how I could update a machine's action, e.g. one that fetches from an API using a token that has refreshed, without forcing a new service (and then having to hydrate the previous service its state).

Just pass it as part of `options` argument, we keep those in sync with the underlying machine:
https://github.com/davidkpiano/xstate/blob/b1684eadb1f859db5c733b8d403afc825c294948/packages/xstate-react/src/useMachine.ts#L234-L236

**Reply by davidkpiano** [reply]
> In similar fashion, I wonder how I could update a machine's action, e.g. one that fetches from an API using a token that has refreshed, _without_ forcing a new service (and then having to hydrate the previous service its state).

This sounds like "notifying" the service of the updated token:

```js
on: {
  UPDATED_TOKEN: {
    actions: forwardTo('theAPIService')
  }
}
```

**Reply by FokkeZB** [reply]
@Andarist - ah, I missed that... awesome!

@davidkpiano that's a nice pattern as well. I think I prefer "patching" actions though, as I'd like to keep the events limited to user events as much as possible.

Thanks both!
